### DISCUSSION TITLE
Array of arrays

### DISCUSSION THREAD

**Discussion by vlc1** [MasonProtter/Bumper.jl#14] [Q&A] [answered]
I'm trying to use Bumper.jl to reduce allocations for arrays of arrays. A typical example is:

```julia
using Bumper
using BenchmarkTools

buf = AllocBuffer(10_000)

# allocating code
function foo(n, m, buf::AllocBuffer, T::Type=Int)
    @no_escape buf begin
        x = Vector{Vector{T}}(undef, n)

        for i in Base.OneTo(n)
            x[i] = alloc(T, buf, m)
        end

        sum(length, x)
    end
end

@btime foo(n, m, buf) setup=(n = 4; m = 32; buf=AllocBuffer(32_768))
```

which allocates as opposed to

```julia
# non-allocating
function bar(n, m, buf::AllocBuffer, ::Type{T}=Int) where {T}
    A = Base._return_type(alloc, Tuple{Type{T},typeof(buf),typeof(m)})

    @no_escape buf begin
        x = alloc(A, buf, n)

        for i in Base.OneTo(n)
            x[i] = alloc(T, buf, m)
        end

        sum(length, x)
    end
end

@btime bar(n, m, buf) setup=(n = 4; m = 32; buf=AllocBuffer(32_768))
```

which doesn't allocate (great!).

My question is: is the way the type `A` in `bar` is inferred correct, or is there a better way to do this?

Thanks!

**Comment by MasonProtter** [comment, marked as answer]
Ah, good question @vlc1, yes this is a bit tricky. You see, the problem is that `alloc` does not return a `Vector` , but instead returns a `PtrArray`:
```julia
julia> @no_escape begin
           typeof(alloc(Int, 10))
       end
PtrArray{Int64, 1, (1,), Tuple{Int64}, Tuple{Nothing}, Tuple{StaticInt{1}}} (alias for StrideArraysCore.AbstractPtrArray{Int64, 1, (1,), Tuple{Int64}, Tuple{Nothing}, Tuple{Static.StaticInt{1}}, Int64})
```
so when you try to put that object into a `Vector{Vector{Int}}`, julia will end up calling `convert` on it which turns it into a regular heap allocated vector!

**Reply by vlc1** [reply]
Thanks @MasonProtter ! That makes sense, I was a bit puzzled by the extra allocations in the `foo` function but your explanation makes sense.

I'll keep on relying on `Base._return_type` then, thanks for your explanation and this great package!
